mathics/core/expression.py

```diff
@@ -943,11 +1060,27 @@ def rules():
             new = Expression(head)
             new._leaves = tuple(dirty_leaves)
 
+        # copy the "unformatted" form of the original expression,
         new.unformatted = self.unformatted
+        # Step 8:updates the cache and returns the new form, with the reevaluate flag to false.
         new._timestamp_cache(evaluation)
```

Comment:
I am not sure why we would need to create a new timestamp rather than keep a copy of the old one and restore that.

The same is true for the code above.  

Some care is needed that when we create new timestamps we aren't sharing information from old once, but I think that's already covered.  That said, at least for the cache, I really am not all that interested in resolving what's up here until we have the rest properly isolated made modular and have tests for these.